In [ ]:
'''
@Time    :   2025/03/01 20:53:39
@Author  :   Adriano Nogueira
'''

In [1]:
!pip install --upgrade openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 472.3/472.3 kB 5.6 MB/s eta 0:00:00
  Attempting uninstall: openai
    Found existing installation: openai 1.61.1
    Uninstalling openai-1.61.1:
      Successfully uninstalled openai-1.61.1


In [2]:
!pip install python-dotenv

# Importing Libraries

In [3]:
from google.colab import drive
from dotenv import load_dotenv
import os
from openai import OpenAI

# Mouting the Drive

In [4]:
drive.mount('/content/drive')

Mounted at /content/drive


# Acessing API Keys

In [5]:
# Change to your environment variable location in your Google Drive
load_dotenv('/content/drive/MyDrive/Unsupervised Machine Learning/BA 820 Team 6/.env')

api_key = os.getenv('API_KEY')

OpenAI.api_key = api_key

# Prompt Engineering Tests

In [ ]:
prompt = "Given a situation where a user is trying to understand the technique of prompt engineering, how would you explain it?"


client = OpenAI(
    api_key=OpenAI.api_key,
)

chat_completion = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": prompt,
        }
    ],
    model="gpt-3.5-turbo",
    max_tokens=150,
    temperature=0.6
)

print(chat_completion.choices[0].message.content.strip())

Prompt engineering is a technique used in artificial intelligence and machine learning to design and create effective prompts that can guide the model to generate desired outputs. It involves carefully crafting the input provided to the model in order to influence the type of responses it produces.

In prompt engineering, the goal is to design prompts that are clear, specific, and tailored to the task at hand. By providing the model with the right context and information, prompt engineering can help improve the accuracy and relevance of the model's outputs.

Some key aspects of prompt engineering include choosing the right language and formatting for the prompt, selecting relevant examples and data to include in the prompt, and fine-tuning the prompt based on the model's performance.

Overall, prompt engineering is an important technique for optimizing


In [ ]:
prompt = "Given a situation where a user is trying to understand the technique of prompt engineering, how would you explain it?"

client = OpenAI(
    api_key=OpenAI.api_key,
)

chat_completion = client.chat.completions.create(
messages = [
    {"role": "system", "content": "You are a postdoctoral researcher in scientific computing."},
    {"role": "user", "content": prompt},
],
    model="gpt-3.5-turbo",
    max_tokens=150,
    temperature=0.6
)

print(chat_completion.choices[0].message.content.strip())

Prompt engineering is a technique used in scientific computing to improve the efficiency and accuracy of simulations or calculations by carefully selecting the initial conditions or prompts given to the system being studied. The goal of prompt engineering is to guide the system towards a desired outcome or behavior by choosing the right inputs that will lead to the desired result.

In practice, prompt engineering involves designing and choosing the prompts or initial conditions in such a way that they encourage the system to evolve in a specific direction or exhibit certain properties. This can be particularly useful in complex systems where the precise outcome is not known in advance, but certain behaviors or characteristics are desired.

By strategically engineering the prompts, researchers can potentially reduce the computational effort needed to reach a particular outcome or improve the accuracy of the results obtained


# Image + Prompt Engineering

In [ ]:
client = OpenAI(
    api_key=OpenAI.api_key,
)

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "Can you describe this chart?"},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://raw.githubusercontent.com/vis-nlp/ChartQA/main/ChartQA%20Dataset/val/png/20294671002019.png",
                    },
                },
            ],
        }
    ],
    max_tokens=300,
)

print(response.choices[0])

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='The chart displays the annual death rate from neonatal preterm birth complications per 100,000 children under 5 years old for various countries, specifically from the year 2004. \n\nKey points from the chart include:\n\n- **Laos** has the highest death rate at **396 deaths** per 100,000 children, indicated by the tallest bar.\n- **Comoros** follows with a significantly lower rate of **180 deaths**.\n- **Nicaragua** is next with **68 deaths**.\n- **Australia** has a much lower rate of **18 deaths**.\n- **Belgium** records the lowest rate in this chart, with **14 deaths**.\n\nThe vertical axis represents the death rates, while each colored bar corresponds to a different country, illustrating stark disparities in neonatal fatality rates associated with preterm birth complications.', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))


In [ ]:
client = OpenAI(
    api_key=OpenAI.api_key,
)

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "Can you produce business insights with the numbers of this chart?"},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://raw.githubusercontent.com/vis-nlp/ChartQA/main/ChartQA%20Dataset/val/png/20294671002019.png",
                    },
                },
            ],
        }
    ],
    max_tokens=300,
)

print(response.choices[0])

Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content="Based on the chart depicting the death rate from neonatal preterm birth complications in 2004:\n\n1. **High Mortality in Laos**: With a death rate of 396 per 100,000 children under five, Laos experiences a significantly higher mortality rate compared to the other countries listed. This indicates a critical need for enhanced maternal and neonatal health care services in Laos.\n\n2. **Moderate Rates in Comoros and Nicaragua**: Comoros has a death rate of 180, while Nicaragua's rate is 68. This suggests that both regions face challenges, but Comoros is much closer in severity to Laos than to Nicaragua, indicating the need for targeted health interventions.\n\n3. **Low Rates in Australia and Belgium**: Australia (18) and Belgium (14) show considerably lower rates of child deaths from this complication. Their health systems likely have better maternal care practices, access to neonatal services, an

# Final Prompt

Your Task:

1) Chart Description (Max 50 words):
Identify the chart type (e.g., bar chart, line graph, pie chart).
Explain what the chart is about (e.g., sales trends, customer demographics).
Summarize the key information it presents in simple terms.

2) Insights & Actions (Max 50 words):
Based on your description (and additional details from the URL), provide actionable insights from the chart.
Include suggestions, next steps, or key takeaways relevant to a small business owner.

Persona:
A U.S.-based small business owner with some college education or below.
Use plain, easy-to-understand English—avoid technical jargon.

In [7]:
client = OpenAI(
    api_key=OpenAI.api_key,
)

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {   "role": "system", "content": "You are an U.S.-based small business owner with some college education or below. Use plain, easy-to-understand English—avoid technical jargon",
            "role": "user",
            "content": [
                {"type": "text", "text": "1) Chart Description (Max 50 words): Identify the chart type (e.g., bar chart, line graph, pie chart). Explain what the chart is about (e.g., sales trends, customer demographics). Summarize the key information it presents in simple terms. 2) Insights & Actions (Max 50 words): Based on your description (and additional details from the URL), provide actionable insights from the chart. Include suggestions, next steps, or key takeaways relevant to a small business owner."},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://raw.githubusercontent.com/vis-nlp/ChartQA/main/ChartQA%20Dataset/val/png/20294671002019.png",
                    },
                },
            ],
        }
    ]
)

print(response.choices[0])

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='1) **Chart Description:** This bar chart illustrates the death rate from neonatal preterm birth complications in 2004, measured as annual child deaths per 100,000 children under five years old. Laos has the highest rate at 396, followed by Comoros (180), Nicaragua (68), Australia (18), and Belgium (14).\n\n2) **Insights & Actions:** The significant disparity in death rates suggests a need for improved maternal and neonatal healthcare in high-risk countries like Laos and Comoros. Small business owners in healthcare could explore opportunities to provide educational resources, training, or medical services aimed at reducing neonatal mortality.', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))


1) **Chart Description:** This bar chart illustrates the death rate from neonatal preterm birth complications in 2004, measured as annual child deaths per 100,000 children under five years old. Laos has the highest rate at 396, followed by Comoros (180), Nicaragua (68), Australia (18), and Belgium (14).

2) **Insights & Actions:** The significant disparity in death rates suggests a need for improved maternal and neonatal healthcare in high-risk countries like Laos and Comoros. Small business owners in healthcare could explore opportunities to provide educational resources, training, or medical services aimed at reducing neonatal mortality.